### PR TITLE
DXCDT-342: Added `AUTH0_INCLUDED_ONLY` configuration property

### PR DESCRIPTION
**Pull request by willvedd**
### 🔧 Changes

This PR introduces the `AUTH0_INCLUDED_ONLY` configuration property which dictates which resources to _only_ manage by the Deploy CLI. This is the opposite to the `AUTH0_EXCLUDED` configuration property which excludes specific resources. Inclusion functionality is helpful when a developer workflow is only concerned with a specific subset of Auth0 resources; this is currently possible through exclusion, but may require an unwieldy list. Some demand for this functionality can be seen in #702.

**Example:**
```json
{
  "AUTH0_INCLUDED_ONLY": [ "connections", "clients", "branding", "tenant" ]
}
```

Further, this functionality had also been pondered in the Resource Exclusion RFC (#468):

> In the future however, it may be possible for this tool to include a AUTH0_INCLUDED property in combination with a wildcard operator available for AUTH0_EXCLUDED.

This feature works as expected, only managing the resource types listed during both import and export. However, there are two key validations:
 - Disallowing its use in conjunction with `AUTH0_EXCLUDED` - both configuration properties are mutually exclusive, using both only introduces complexity
 - Disallowing an empty value - only including zero resources is a no-op

### 📚 References

- Resource Exclusion Proposal (#468)
- Feature request Github Issue (#702)

### 🔬 Testing

Lots of testing with this one, with six unit tests and one end-to-end test. Also manual testing on-top.

### 📝 Checklist

- [x] All new/changed/fixed functionality is covered by tests (or N/A)
- [x] I have added documentation for all new/changed functionality (or N/A)

### CODE REVIEW

File: test/tools/auth0/index.test.ts [hunk 1]
```diff
@@ -8,10 +8,16 @@ const mockEmptyAssets = {} as Assets;
...
+        const config = { AUTH0_EXCLUDED: [] };
+        return config[key];
+      });
 
       const AUTH0_EXCLUDED = ['rules', 'organizations', 'connections'];
-      const auth0WithExclusions = new Auth0(mockEmptyClient, mockEmptyAssets, () => AUTH0_EXCLUDED);
+      const auth0WithExclusions = new Auth0(mockEmptyClient, mockEmptyAssets, (key) => {
+        const config = { AUTH0_EXCLUDED };
+        return config[key];
+      });
```
Comment (willvedd) [inline test/tools/auth0/index.test.ts:20]: The configuration function injected into these tests were faulty and caused false-positives when the `AUTH0_INCLUDED_ONLY` property was added.

File: docs/configuring-the-deploy-cli.md [hunk 2]
```diff
@@ -82,6 +84,20 @@ Array of strings. Excludes entire resource types from being managed, bi-directio
...
+
+#### Example
+
+```json
+{
+  "AUTH0_INCLUDED_ONLY": ["clients", "connections", "tenant", "branding"]
+}
+```
+
+Cannot be used simultaneously with `AUTH0_INCLUDED_ONLY`.
```
Comment (willvedd) [inline docs/configuring-the-deploy-cli.md:99]:
```suggestion
Cannot be used simultaneously with `AUTH0_EXCLUDED`.
```

File: src/context/index.ts [hunk 3]
```diff
@@ -128,3 +150,10 @@ export const setupContext = async (config: Config): Promise<DirectoryContext | Y
     `Unable to determine context processor to load for file ${inputFile}, does it exist? `
   );
 };
+
+export const filterOnlyIncludedResourceTypes =
```
Comment (willvedd) [inline src/context/index.ts:154]: Debatably overkill to pull into its own function, but this filtering function was being used in several foundational areas of code. I think that encapsulating the logic of it is important. Also, it enables additional unit testing.

File: src/context/index.ts [hunk 4]
```diff
@@ -40,6 +42,26 @@ export const setupContext = async (config: Config): Promise<DirectoryContext | Y
     );
   }
 
+  ((config: Config) => {
+    if (config.AUTH0_INCLUDED_ONLY === undefined) return;
+
+    if (config.AUTH0_INCLUDED_ONLY.length === 0) {
+      throw new Error(
+        `Need to define at least one resource type in AUTH0_INCLUDED_ONLY configuration.`
```
Comment (sergiught) [inline src/context/index.ts:50]: Wdyt about including a link to the docs? e.g. "For more info: https://github.com/auth0/auth0-deploy-cli/blob/master/docs/configuring-the-deploy-cli.md"

File: src/context/index.ts [hunk 5]
```diff
@@ -40,6 +42,26 @@ export const setupContext = async (config: Config): Promise<DirectoryContext | Y
...
+      );
+    }
+
+    const hasExcludedResources =
+      config.AUTH0_EXCLUDED !== undefined && config.AUTH0_EXCLUDED.length > 0;
+    const hasIncludedResources = config.AUTH0_INCLUDED_ONLY.length > 0;
+
+    if (hasExcludedResources && hasIncludedResources) {
+      throw new Error(
+        `Both AUTH0_EXCLUDED and AUTH0_INCLUDED_ONLY configuration values are defined`
```
Comment (sergiught) [inline src/context/index.ts:60]:
```suggestion
        `Both AUTH0_EXCLUDED and AUTH0_INCLUDED_ONLY configuration values are defined. You can only make use of one. For more info... <link to docs>.` 
```

wdyt?

**Review by sergiught — APPROVED** [review]
Good work 👍🏻 

I tested this locally as well with an include of only `clients` and works as intended. 

```shell
$ node lib/index.js export -c config.json -f directory  -o ./manual-testing/directory
2023-01-31T13:00:15.733Z - info: Loading Auth0 Tenant Data
2023-01-31T13:00:16.044Z - info: Retrieving clients data from Auth0
2023-01-31T13:00:16.048Z - info: Writing manual-testing/directory/clients/Regular Web Application For Quickstarts.json
2023-01-31T13:00:16.049Z - info: Writing manual-testing/directory/clients/API Explorer Application.json
2023-01-31T13:00:16.050Z - info: Writing manual-testing/directory/clients/Terraform Provider Auth0.json
2023-01-31T13:00:16.050Z - info: Writing manual-testing/directory/clients/Deploy CLI.json
2023-01-31T13:00:16.051Z - info: Writing manual-testing/directory/clients/CLI Login Testing.json
2023-01-31T13:00:16.052Z - info: Export Successful
```

Only left 2 suggestions to improve the error message, but it's not a blocker.

File: src/context/index.ts (hunk 5 above)
Comment (willvedd) [reply src/context/index.ts:60]: That is a good idea, will implement.

File: src/context/index.ts (hunk 4 above)
Comment (willvedd) [reply src/context/index.ts:50]: Makes total sense, will add!

**Pull request merged**